src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs

```diff
@@ -441,6 +443,212 @@ public List<PackageManagerSearchElement> ListAll()
             return CachedPackageList;
         }
 
+        internal void InitiatePackageDownloadAndInstall(IPackageInfo packageInfo, string downloadPath = null)
+        {
+            var header = Model.GetPackageHeader(packageInfo);
+            if (header == null)
+            {
+                // TODO: move message to resources
+                var result = MessageBox.Show(string.Format("{0} could not be downloaded.", packageInfo.Name),
+                "Download Error", MessageBoxButton.OKCancel, MessageBoxImage.Question);
+                return;
+            }
+
+            var version = Model.GetGregPackageVersion(header, packageInfo.Version);
+            if (version == null)
+            {
+                // TODO: move message to resources
+                var result = MessageBox.Show(string.Format("Version {0} of {1} could not be found.", packageInfo.Version.ToString(), packageInfo.Name),
+                "Download Error", MessageBoxButton.OKCancel, MessageBoxImage.Question);
+                return;
+            }
+
+            ExecutePackageDownload(packageInfo.Name, version, downloadPath);
+        }
+
+        private string JoinPackageNames(IEnumerable<Package> pkgs)
+        {
+            return String.Join(", ", pkgs.Select(x => x.Name + " " + x.VersionName));
+        }
+
+        internal void ExecutePackageDownload(string name, PackageVersion version, string downloadPath)
+        {
+            string msg = String.IsNullOrEmpty(downloadPath) ?
+                String.Format(Resources.MessageConfirmToInstallPackage, name, version.version) :
+                String.Format(Resources.MessageConfirmToInstallPackageToFolder, name, version.version, downloadPath);
+
+            var result = MessageBox.Show(msg,
+                Resources.PackageDownloadConfirmMessageBoxTitle,
+                MessageBoxButton.OKCancel, MessageBoxImage.Question);
+
+            var pmExt = DynamoViewModel.Model.GetPackageManagerExtension();
+            if (result == MessageBoxResult.OK)
+            {
+                // get all of the headers
+                var headers = version.full_dependency_ids.Select(dep => dep._id).Select((id) =>
+                {
+                    PackageHeader pkgHeader;
+                    var res = pmExt.PackageManagerClient.DownloadPackageHeader(id, out pkgHeader);
```

Comment:
Since you are already in the PackageManagerClientViewModel, I think you no longer need to access all the way from extension, `Model` here should be client right?